File: android/Reguerta/app/src/main/java/com/reguerta/user/data/access/FirestoreMemberRepository.kt
package com.reguerta.user.data.access

import com.google.firebase.firestore.FirebaseFirestore
import com.reguerta.user.data.firestore.ReguertaFirestoreCollection
import com.reguerta.user.data.firestore.ReguertaFirestoreEnvironment
import com.reguerta.user.data.firestore.ReguertaFirestorePath
import com.reguerta.user.data.firestore.toRepositoryException
import com.reguerta.user.domain.RepositoryErrorKind
import com.reguerta.user.domain.RepositoryException
import com.reguerta.user.domain.access.EcoCommitmentMode
import com.reguerta.user.domain.access.Member
import com.reguerta.user.domain.access.MemberRepository
import com.reguerta.user.domain.access.MemberRole
import com.reguerta.user.domain.access.ProducerParity
import com.reguerta.user.domain.access.canManageMembers
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.tasks.await
import kotlinx.coroutines.withContext

class FirestoreMemberRepository(
    private val firestore: FirebaseFirestore,
    private val environment: ReguertaFirestoreEnvironment? = null,
) : MemberRepository {
    private val firestorePath = ReguertaFirestorePath(environment = environment)

    private val usersCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.USERS)

    private val authLinksCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.AUTH_LINKS)

    private val memberDirectoryCollectionPath: String
        get() = firestorePath.collectionPath(ReguertaFirestoreCollection.MEMBER_DIRECTORY)

    override suspend fun findByAuthUid(authUid: String): Member? = withContext(Dispatchers.IO) {
        try {
            val authLink = firestore.collection(authLinksCollectionPath)
                .document(authUid)
                .get()
                .await()
            if (!authLink.exists()) return@withContext null
            val authLinkData: Map<String, Any> = authLink.data ?: invalidDocument("authLinks.document")
            val memberId = authLinkData.requiredString(
                keys = arrayOf("memberId"),
                resource = "authLinks.document",
            )
            val memberSnapshot = firestore.collection(usersCollectionPath)
                .document(memberId)
                .get()
                .await()
            if (!memberSnapshot.exists()) {
                null
            } else {
                val memberData: Map<String, Any> = memberSnapshot.data ?: invalidDocument("members.document")
                decodeMemberDocument(memberSnapshot.id, memberData)
            }
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "members.authLink")
        }
    }

    override suspend fun getMembersVisibleTo(member: Member): List<Member> = withContext(Dispatchers.IO) {
        try {
            val canReadPrivateMembers = member.canManageMembers
            val snapshot = if (canReadPrivateMembers) {
                firestore.collection(usersCollectionPath).get()
            } else {
                firestore.collection(memberDirectoryCollectionPath)
                    .whereEqualTo("isActive", true)
                    .get()
            }.await()
            val visible = snapshot.documents.map { document ->
                val data: Map<String, Any> = document.data ?: invalidDocument(
                    if (canReadPrivateMembers) "members.document" else "members.directory.document",
                )
                if (canReadPrivateMembers) {
                    decodeMemberDocument(document.id, data)
                } else {
                    decodeDirectoryMemberDocument(document.id, data)
                }
            }.map { candidate ->
                if (candidate.id == member.id) member else candidate
            }.toMutableList()
            if (visible.none { it.id == member.id }) {
                visible += member
            }
            visible.sortedBy { it.displayName.lowercase() }
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "members")
        }
    }

    override suspend fun updateOwnProducerCatalogEnabled(
        member: Member,
        isEnabled: Boolean,
    ): Member = withContext(Dispatchers.IO) {
        try {
            firestore.collection(usersCollectionPath)
                .document(member.id)
                .update(mapOf("producerCatalogEnabled" to isEnabled))
                .await()
            member.copy(producerCatalogEnabled = isEnabled)
        } catch (error: Exception) {
            throw error.toRepositoryException(resource = "members.write")
        }
    }
}

internal fun decodeDirectoryMemberDocument(
    documentId: String,
    data: Map<String, Any?>,
): Member {
    val resource = "members.directory.document"
    val id = requiredDocumentId(documentId, resource)
    if (data.requiredString(arrayOf("userId"), resource) != id) invalidDocument(resource)
    val displayName = data.requiredString(arrayOf("displayName"), resource)
    val companyName = data.optionalString(arrayOf("companyName"), resource)
    val roles = data.directoryRoles(resource)
    if (!data.requiredBoolean("isActive", resource)) invalidDocument(resource)
    val producerCatalogEnabled = data.requiredBoolean("producerCatalogEnabled", resource)
    val isCommonPurchaseManager = data.requiredBoolean("isCommonPurchaseManager", resource)
    val producerParity = optionalParity(data["producerParity"], resource, acceptsLegacyCasing = false)
    val ecoCommitment = data.optionalMap("ecoCommitment", resource) ?: invalidDocument(resource)
    val ecoCommitmentMode = requiredEcoMode(ecoCommitment["mode"], resource)
    val ecoCommitmentParity = optionalParity(
        ecoCommitment["parity"],
        resource,
        acceptsLegacyCasing = false,
    )

    return Member(
        id = id,
        displayName = displayName,
        companyName = companyName,
        phoneNumber = null,
        normalizedEmail = "",
        authUid = null,
        roles = roles,
        isActive = true,
        producerCatalogEnabled = producerCatalogEnabled,
        isCommonPurchaseManager = isCommonPurchaseManager,
        producerParity = producerParity,
        ecoCommitmentMode = ecoCommitmentMode,
        ecoCommitmentParity = ecoCommitmentParity,
    )
}

internal fun decodeMemberDocument(
    documentId: String,
    data: Map<String, Any?>,
): Member {
    val resource = "members.document"
    val id = requiredDocumentId(documentId, resource)
    val displayName = data.optionalString(arrayOf("displayName"), resource)
        ?: listOfNotNull(
            data.optionalString(arrayOf("name"), resource),
            data.optionalString(arrayOf("surname"), resource),
        ).joinToString(" ").ifBlank { null }
        ?: invalidDocument(resource)
    val normalizedEmail = data.optionalString(
        arrayOf("normalizedEmail", "emailNormalized", "email"),
        resource,
    )?.lowercase() ?: invalidDocument(resource)
    val authUid = data.optionalString(arrayOf("authUid"), resource)
    val companyName = data.optionalString(arrayOf("companyName", "company_name", "company"), resource)
    val phoneNumber = data.optionalString(
        arrayOf("phoneNumber", "phone", "telephone", "telefono"),
        resource,
    )
    val isActive = data.optionalBoolean(arrayOf("isActive", "available"), true, resource)
    val producerCatalogEnabled = data.optionalBoolean(arrayOf("producerCatalogEnabled"), true, resource)
    val isCommonPurchaseManager = data.optionalBoolean(
        arrayOf("isCommonPurchaseManager"),
        false,
        resource,
    )
    val producerParity = optionalParity(data["producerParity"], resource, acceptsLegacyCasing = true)
    val ecoCommitment = data.optionalMap("ecoCommitment", resource)
    val ecoCommitmentMode = optionalEcoMode(
        ecoCommitment?.get("mode"),
        EcoCommitmentMode.WEEKLY,
        resource,
        acceptsLegacyCasing = true,
    )
    val ecoCommitmentParity = optionalParity(
        ecoCommitment?.get("parity"),
        resource,
        acceptsLegacyCasing = true,
    )
    val roles = data.fullMemberRoles(resource)

    return Member(
        id = id,
        displayName = displayName,
        companyName = companyName,
        phoneNumber = phoneNumber,
        normalizedEmail = normalizedEmail,
        authUid = authUid,
        roles = roles,
        isActive = isActive,
        producerCatalogEnabled = producerCatalogEnabled,
        isCommonPurchaseManager = isCommonPurchaseManager,
        producerParity = producerParity,
        ecoCommitmentMode = ecoCommitmentMode,
        ecoCommitmentParity = ecoCommitmentParity,
    )
}

private fun requiredDocumentId(documentId: String, resource: String): String {
    val normalized = documentId.trim()
    if (normalized.isBlank() || normalized.contains('/')) invalidDocument(resource)
    return normalized
}

private fun Map<String, Any?>.requiredString(keys: Array<String>, resource: String): String =
    optionalString(keys, resource) ?: invalidDocument(resource)

private fun Map<String, Any?>.optionalString(keys: Array<String>, resource: String): String? {
    keys.forEach { key ->
        val rawValue = this[key] ?: return@forEach
        val value = rawValue as? String ?: invalidDocument(resource)
        value.trim().ifBlank { null }?.let { return it }
    }
    return null
}

private fun Map<String, Any?>.requiredBoolean(field: String, resource: String): Boolean =
    this[field] as? Boolean ?: invalidDocument(resource)

private fun Map<String, Any?>.optionalBoolean(
    keys: Array<String>,
    default: Boolean,
    resource: String,
): Boolean {
    keys.forEach { key ->
        val rawValue = this[key] ?: return@forEach
        return rawValue as? Boolean ?: invalidDocument(resource)
    }
    return default
}

private fun Map<String, Any?>.optionalMap(field: String, resource: String): Map<String, Any?>? {
    val rawValue = this[field] ?: return null
    @Suppress("UNCHECKED_CAST")
    return rawValue as? Map<String, Any?> ?: invalidDocument(resource)
}

private fun Map<String, Any?>.fullMemberRoles(resource: String): Set<MemberRole> {
    val rawRoles = this["roles"]
    val parsedRoles = if (rawRoles == null) {
        emptySet()
    } else {
        val values = rawRoles as? List<*> ?: invalidDocument(resource)
        values.map { value ->
            (value as? String)?.toLegacyMemberRoleOrNull() ?: invalidDocument(resource)
        }.toSet()
    }
    if (parsedRoles.isNotEmpty()) return parsedRoles
    return buildSet {
        add(MemberRole.MEMBER)
        if (optionalBoolean(arrayOf("isProducer"), false, resource)) add(MemberRole.PRODUCER)
        if (optionalBoolean(arrayOf("isAdmin"), false, resource)) add(MemberRole.ADMIN)
    }
}

private fun Map<String, Any?>.directoryRoles(resource: String): Set<MemberRole> {
    val values = this["roles"] as? List<*> ?: invalidDocument(resource)
    if (values.isEmpty()) invalidDocument(resource)
    val roles = values.map { value ->
        when (value) {
            "member" -> MemberRole.MEMBER
            "producer" -> MemberRole.PRODUCER
            "admin" -> MemberRole.ADMIN
            else -> invalidDocument(resource)
        }
    }.toSet()
    if (!roles.contains(MemberRole.MEMBER)) invalidDocument(resource)
    return roles
}

private fun optionalParity(
    rawValue: Any?,
    resource: String,
    acceptsLegacyCasing: Boolean,
): ProducerParity? {
    val value = rawValue ?: return null
    val string = value as? String ?: invalidDocument(resource)
    val normalized = string.trim().let { if (acceptsLegacyCasing) it.lowercase() else it }
    return when (normalized) {
        "even" -> ProducerParity.EVEN
        "odd" -> ProducerParity.ODD
        else -> invalidDocument(resource)
    }
}

private fun requiredEcoMode(rawValue: Any?, resource: String): EcoCommitmentMode {
    val value = rawValue as? String ?: invalidDocument(resource)
    if (value != value.trim()) invalidDocument(resource)
    return when (value) {
        "weekly" -> EcoCommitmentMode.WEEKLY
        "biweekly" -> EcoCommitmentMode.BIWEEKLY
        else -> invalidDocument(resource)
    }
}

private fun optionalEcoMode(
    rawValue: Any?,
    default: EcoCommitmentMode,
    resource: String,
    acceptsLegacyCasing: Boolean,
): EcoCommitmentMode {
    val value = rawValue ?: return default
    val string = value as? String ?: invalidDocument(resource)
    val normalized = string.trim().let { if (acceptsLegacyCasing) it.lowercase() else it }
    return when (normalized) {
        "weekly" -> EcoCommitmentMode.WEEKLY
        "biweekly" -> EcoCommitmentMode.BIWEEKLY
        else -> invalidDocument(resource)
    }
}

private fun String.toLegacyMemberRoleOrNull(): MemberRole? = when (trim().lowercase()) {
    "member", "socio" -> MemberRole.MEMBER
    "producer", "productor" -> MemberRole.PRODUCER
    "admin", "administrador" -> MemberRole.ADMIN
    else -> null
}

private fun invalidDocument(resource: String): Nothing = throw RepositoryException(
    kind = RepositoryErrorKind.INVALID_DATA,
    resource = resource,
)

File: ios/Reguerta/Reguerta/Data/Access/FirestoreMemberRepository.swift
import FirebaseCore
import FirebaseFirestore
import Foundation

actor FirestoreMemberRepository: MemberRepository {
    private let storedDB: Firestore

    init(firebaseAppName: String) {
        guard let app = FirebaseApp.app(name: firebaseAppName) else {
            preconditionFailure("Firebase app is required for members")
        }
        self.storedDB = Firestore.firestore(app: app)
    }

    func member(id: String, environment: SessionEnvironment) async throws -> Member? {
        let normalizedID = id.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalizedID.isEmpty, !normalizedID.contains("/") else {
            throw RepositoryError.invalidData(resource: "members.id")
        }

        do {
            try Task.checkCancellation()
            let usersCollection = storedDB.reguertaCollection(.users, environment: environment)
            let snapshot = try await usersCollection.document(normalizedID).getDocument()
            guard snapshot.exists else { return nil }
            guard let data = snapshot.data() else { throw Self.invalidMemberDocumentError }
            return try Self.member(documentID: snapshot.documentID, data: data)
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "members.document")
        }
    }

    func members(visibleTo member: Member, environment: SessionEnvironment) async throws -> [Member] {
        do {
            try Task.checkCancellation()
            let members: [Member]
            if member.canManageMembers {
                let usersCollection = storedDB.reguertaCollection(.users, environment: environment)
                let snapshot = try await usersCollection.getDocuments()
                members = try snapshot.documents.map { document in
                    try Self.member(documentID: document.documentID, data: document.data())
                }
            } else {
                let memberDirectoryCollection = storedDB.reguertaCollection(
                    .memberDirectory,
                    environment: environment
                )
                let snapshot = try await memberDirectoryCollection
                    .whereField("isActive", isEqualTo: true)
                    .getDocuments()
                let directoryMembers = try snapshot.documents.map { document in
                    try Self.directoryMember(documentID: document.documentID, data: document.data())
                }
                members = Self.mergingAuthenticatedMember(member, into: directoryMembers)
            }
            return members.sorted { lhs, rhs in
                lhs.displayName.localizedCaseInsensitiveCompare(rhs.displayName) == .orderedAscending
            }
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "members")
        }
    }

    func updateOwnProducerCatalogEnabled(
        member: Member,
        enabled: Bool,
        environment: SessionEnvironment
    ) async throws -> Member {
        do {
            try Task.checkCancellation()
            let usersCollection = storedDB.reguertaCollection(.users, environment: environment)
            try await usersCollection.document(member.id).updateData([
                "producerCatalogEnabled": enabled
            ])
        } catch {
            throw FirestoreRepositoryErrorMapper.map(error, resource: "members.write")
        }
        return member.copy(producerCatalogEnabled: enabled)
    }
}

extension FirestoreMemberRepository {
    static func member(documentID: String, data: [String: Any]) throws -> Member {
        let id = try requiredDocumentID(documentID, resource: "members.document")
        let contact = try fullMemberContact(data)
        let access = try fullMemberAccess(data)
        let commitment = try fullMemberCommitment(data)

        return Member(
            id: id,
            displayName: contact.displayName,
            companyName: contact.companyName,
            phoneNumber: contact.phoneNumber,
            normalizedEmail: contact.normalizedEmail,
            authUid: contact.authUid,
            roles: access.roles,
            isActive: access.isActive,
            producerCatalogEnabled: access.producerCatalogEnabled,
            isCommonPurchaseManager: access.isCommonPurchaseManager,
            producerParity: commitment.producerParity,
            ecoCommitmentMode: commitment.mode,
            ecoCommitmentParity: commitment.parity
        )
    }

    static func directoryMember(documentID: String, data: [String: Any]) throws -> Member {
        let resource = "members.directory.document"
        let id = try requiredDocumentID(documentID, resource: resource)
        guard try requiredString(data, field: "userId", resource: resource) == id else {
            throw RepositoryError.invalidData(resource: resource)
        }
        let displayName = try requiredString(data, field: "displayName", resource: resource)
        let companyName = try optionalString(data, keys: ["companyName"], resource: resource)
        let roles = try directoryRoles(data, resource: resource)
        guard try requiredBool(data, field: "isActive", resource: resource) else {
            throw RepositoryError.invalidData(resource: resource)
        }
        let producerCatalogEnabled = try requiredBool(
            data,
            field: "producerCatalogEnabled",
            resource: resource
        )
        let isCommonPurchaseManager = try requiredBool(
            data,
            field: "isCommonPurchaseManager",
            resource: resource
        )
        let producerParity = try optionalParity(
            data["producerParity"],
            resource: resource,
            acceptsLegacyValues: false
        )
        guard let ecoCommitment = try optionalMap(data, field: "ecoCommitment", resource: resource) else {
            throw RepositoryError.invalidData(resource: resource)
        }
        let ecoCommitmentMode = try requiredEcoMode(ecoCommitment["mode"], resource: resource)
        let ecoCommitmentParity = try optionalParity(
            ecoCommitment["parity"],
            resource: resource,
            acceptsLegacyValues: false
        )

        return Member(
            id: id,
            displayName: displayName,
            companyName: companyName,
            phoneNumber: nil,
            normalizedEmail: "",
            authUid: nil,
            roles: roles,
            isActive: true,
            producerCatalogEnabled: producerCatalogEnabled,
            isCommonPurchaseManager: isCommonPurchaseManager,
            producerParity: producerParity,
            ecoCommitmentMode: ecoCommitmentMode,
            ecoCommitmentParity: ecoCommitmentParity
        )
    }

    static func mergingAuthenticatedMember(_ authenticatedMember: Member, into directoryMembers: [Member]) -> [Member] {
        directoryMembers.filter { $0.id != authenticatedMember.id } + [authenticatedMember]
    }
}

private extension FirestoreMemberRepository {
    static func fullMemberContact(_ data: [String: Any]) throws -> FullMemberContact {
        let resource = "members.document"
        let displayName = try optionalString(data, keys: ["displayName"], resource: resource)
            ?? combinedName(
                firstName: try optionalString(data, keys: ["name"], resource: resource),
                lastName: try optionalString(data, keys: ["surname"], resource: resource)
            )
        guard let displayName,
              let normalizedEmail = try optionalString(
                data,
                keys: ["normalizedEmail", "emailNormalized", "email"],
                resource: resource
              )?.lowercased() else {
            throw invalidMemberDocumentError
        }
        return FullMemberContact(
            displayName: displayName,
            companyName: try optionalString(
                data,
                keys: ["companyName", "company_name", "company"],
                resource: resource
            ),
            phoneNumber: try optionalString(
                data,
                keys: ["phoneNumber", "phone", "telephone", "telefono"],
                resource: resource
            ),
            normalizedEmail: normalizedEmail,
            authUid: try optionalString(data, keys: ["authUid"], resource: resource)
        )
    }

    static func fullMemberAccess(_ data: [String: Any]) throws -> FullMemberAccess {
        let resource = "members.document"
        return FullMemberAccess(
            roles: try fullMemberRoles(data),
            isActive: try optionalBool(
                data,
                keys: ["isActive", "available"],
                default: true,
                resource: resource
            ),
            producerCatalogEnabled: try optionalBool(
                data,
                keys: ["producerCatalogEnabled"],
                default: true,
                resource: resource
            ),
            isCommonPurchaseManager: try optionalBool(
                data,
                keys: ["isCommonPurchaseManager"],
                default: false,
                resource: resource
            )
        )
    }

    static func fullMemberCommitment(_ data: [String: Any]) throws -> FullMemberCommitment {
        let resource = "members.document"
        let ecoCommitment = try optionalMap(data, field: "ecoCommitment", resource: resource)
        return FullMemberCommitment(
            producerParity: try optionalParity(
                data["producerParity"],
                resource: resource,
                acceptsLegacyValues: true
            ),
            mode: try optionalEcoMode(
                ecoCommitment?["mode"],
                default: .weekly,
                resource: resource,
                acceptsLegacyValues: true
            ),
            parity: try optionalParity(
                ecoCommitment?["parity"],
                resource: resource,
                acceptsLegacyValues: true
            )
        )
    }

    private static func fullMemberRoles(_ data: [String: Any]) throws -> Set<MemberRole> {
        let resource = "members.document"
        let parsedRoles: Set<MemberRole>
        if let rawRoles = data["roles"], !(rawRoles is NSNull) {
            guard let values = rawRoles as? [Any] else { throw RepositoryError.invalidData(resource: resource) }
            parsedRoles = try Set(values.map { value in
                guard let value = value as? String,
                      let role = legacyCompatibleRole(from: value) else {
                    throw RepositoryError.invalidData(resource: resource)
                }
                return role
            })
        } else {
            parsedRoles = []
        }
        if !parsedRoles.isEmpty { return parsedRoles }
        return legacyRoles(
            isProducer: try optionalBool(
                data,
                keys: ["isProducer"],
                default: false,
                resource: resource
            ),
            isAdmin: try optionalBool(
                data,
                keys: ["isAdmin"],
                default: false,
                resource: resource
            )
        )
    }

    private static func directoryRoles(_ data: [String: Any], resource: String) throws -> Set<MemberRole> {
        guard let rawRoles = data["roles"] as? [Any], !rawRoles.isEmpty else {
            throw RepositoryError.invalidData(resource: resource)
        }
        let roles = try Set(rawRoles.map { value in
            guard let value = value as? String,
                  value == value.trimmingCharacters(in: .whitespacesAndNewlines),
                  let role = MemberRole(rawValue: value) else {
                throw RepositoryError.invalidData(resource: resource)
            }
            return role
        })
        guard roles.contains(.member) else { throw RepositoryError.invalidData(resource: resource) }
        return roles
    }

    private static func requiredDocumentID(_ documentID: String, resource: String) throws -> String {
        let normalized = documentID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !normalized.isEmpty, !normalized.contains("/") else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return normalized
    }

    private static func requiredString(_ data: [String: Any], field: String, resource: String) throws -> String {
        guard let value = try optionalString(data, keys: [field], resource: resource) else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return value
    }

    private static func optionalString(_ data: [String: Any], keys: [String], resource: String) throws -> String? {
        for key in keys {
            guard let rawValue = data[key], !(rawValue is NSNull) else { continue }
            guard let value = rawValue as? String else { throw RepositoryError.invalidData(resource: resource) }
            let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
            if !normalized.isEmpty { return normalized }
        }
        return nil
    }

    private static func requiredBool(_ data: [String: Any], field: String, resource: String) throws -> Bool {
        guard let rawValue = data[field], let value = rawValue as? Bool else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return value
    }

    private static func optionalBool(
        _ data: [String: Any],
        keys: [String],
        default defaultValue: Bool,
        resource: String
    ) throws -> Bool {
        for key in keys {
            guard let rawValue = data[key], !(rawValue is NSNull) else { continue }
            guard let value = rawValue as? Bool else { throw RepositoryError.invalidData(resource: resource) }
            return value
        }
        return defaultValue
    }

    private static func optionalMap(_ data: [String: Any], field: String, resource: String) throws -> [String: Any]? {
        guard let rawValue = data[field], !(rawValue is NSNull) else { return nil }
        guard let value = rawValue as? [String: Any] else { throw RepositoryError.invalidData(resource: resource) }
        return value
    }

    private static func optionalParity(
        _ rawValue: Any?,
        resource: String,
        acceptsLegacyValues: Bool
    ) throws -> ProducerParity? {
        guard let rawValue, !(rawValue is NSNull) else { return nil }
        guard let value = rawValue as? String else { throw RepositoryError.invalidData(resource: resource) }
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        let candidate = acceptsLegacyValues ? normalized.lowercased() : normalized
        if acceptsLegacyValues, candidate.isEmpty { return nil }
        if let parity = ProducerParity(rawValue: candidate) {
            return parity
        }
        if acceptsLegacyValues {
            switch candidate {
            case "par": return .even
            case "impar": return .odd
            default: break
            }
        }
        throw RepositoryError.invalidData(resource: resource)
    }

    private static func requiredEcoMode(_ rawValue: Any?, resource: String) throws -> EcoCommitmentMode {
        guard let rawValue, let value = rawValue as? String,
              value == value.trimmingCharacters(in: .whitespacesAndNewlines),
              let mode = EcoCommitmentMode(rawValue: value) else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return mode
    }

    private static func optionalEcoMode(
        _ rawValue: Any?,
        default defaultValue: EcoCommitmentMode,
        resource: String,
        acceptsLegacyValues: Bool
    ) throws -> EcoCommitmentMode {
        guard let rawValue, !(rawValue is NSNull) else { return defaultValue }
        guard let value = rawValue as? String else { throw RepositoryError.invalidData(resource: resource) }
        let normalized = value.trimmingCharacters(in: .whitespacesAndNewlines)
        let candidate = acceptsLegacyValues ? normalized.lowercased() : normalized
        if acceptsLegacyValues, candidate.isEmpty { return defaultValue }
        guard let mode = EcoCommitmentMode(rawValue: candidate) else {
            throw RepositoryError.invalidData(resource: resource)
        }
        return mode
    }

    private static func legacyCompatibleRole(from rawValue: String) -> MemberRole? {
        switch rawValue.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "member", "socio": .member
        case "producer", "productor": .producer
        case "admin", "administrador": .admin
        default: nil
        }
    }

    private static func legacyRoles(isProducer: Bool, isAdmin: Bool) -> Set<MemberRole> {
        var roles: Set<MemberRole> = [.member]
        if isProducer { roles.insert(.producer) }
        if isAdmin { roles.insert(.admin) }
        return roles
    }

    private static func combinedName(firstName: String?, lastName: String?) -> String? {
        let combined = [firstName, lastName].compactMap { $0 }.joined(separator: " ")
        return combined.isEmpty ? nil : combined
    }

    private static var invalidMemberDocumentError: RepositoryError {
        .invalidData(resource: "members.document")
    }
}

private struct FullMemberContact {
    let displayName: String
    let companyName: String?
    let phoneNumber: String?
    let normalizedEmail: String
    let authUid: String?
}

private struct FullMemberAccess {
    let roles: Set<MemberRole>
    let isActive: Bool
    let producerCatalogEnabled: Bool
    let isCommonPurchaseManager: Bool
}

private struct FullMemberCommitment {
    let producerParity: ProducerParity?
    let mode: EcoCommitmentMode
    let parity: ProducerParity?
}
